Signal/src/ViewControllers/ConversationView/ConversationViewLayout.h

```diff
@@ -0,0 +1,38 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, ConversationViewLayoutAlignment) {
+    ConversationViewLayoutAlignment_Incoming,
+    ConversationViewLayoutAlignment_Outgoing,
+    ConversationViewLayoutAlignment_FullWidth,
+    ConversationViewLayoutAlignment_Center,
+};
+
+@protocol ConversationViewLayoutItem <NSObject>
+
+- (CGSize)cellSizeForViewWidth:(int)viewWidth contentWidth:(int)contentWidth;
+
+- (ConversationViewLayoutAlignment)layoutAlignment;
+
+@end
+
+#pragma mark -
+
+@protocol ConversationViewLayoutDelegate <NSObject>
+
+- (NSArray<id<ConversationViewLayoutItem>> *)layoutItems;
+
+@end
+
+#pragma mark -
+
+@interface ConversationViewLayout : UICollectionViewLayout
```

Comment:
A new lean and efficient layout for conversation view designed to handle our edge cases (e.g. full-width unread indicators, etc.).